### ISSUE TITLE
MDC and NioEventLoopGroup

### ISSUE THREAD

**Issue by RaviH** [lettuce-io/lettuce-core#951]
Hi,

I am using lettuce-core `5.1.0.RELEASE`. I have information in the MDC context. When I use the synchronous commands / execution, the log statements in my redis library print the information in the MDC context. Whereas when I use asynchronous context / commands, the log statements don't print MDC info. I am guessing because of different thread groups when using asynchronous context. How can I pass the MDC info to asynchronous context / NioEventLoopGroup thread?

Thanks in advance.

**Comment by mp911de** [comment]
## TL; DR:

There's no good solution to this problem.

## Explanation

The MDC is bound to `ThreadLocal`. If you use async components and you await completion in the original thread, then you stay within the same context. If you want to use MDC in an asynchronous context, then you run into trouble as the async component callback happens on a different thread that is no longer associated with your original MDC. You need MDC propagation and cleanup in that case. 

There are a couple of approaches but the most safe and sane approach is carrying over a MDC context copy to your callback so you can prepare and cleanup the MDC before/after your logging statements.

See also:

* https://github.com/netty/netty/issues/6810
* https://github.com/playframework/playframework/issues/7458

**Comment by RaviH** [comment]
Hi @mp911de ,

It would be awesome if you can provide an example of: 

> most safe and sane approach is carrying over a MDC context copy to your callback so you can prepare and cleanup the MDC before/after your logging statements

Also, netty/netty#6810, I couldn't find a place where I could plug it in lettuce. Any ideas/pointers? Thanks!

**Comment by mp911de** [comment]
The code should resemble something like:

```java
StatefulRedisConnection<String, String> connection = …;

Map<String, String> mdc = MDC.getCopyOfContextMap();
connection.async().set("foo", "bar").whenComplete((status, err) -> {

    MDC.setContextMap(mdc);
    
    try {

        LoggerFactory.getLogger("some.logger").info(…);
        
    }   finally {
        MDC.setContextMap(new HashMap<>());
    }
    
});
```
